### DISCUSSION TITLE
How to fake GPS and Compass via MSP

### DISCUSSION THREAD

**Discussion by yannbouteiller** [iNavFlight/inav#8516] [Q&A]
Hello,

We are currently trying to fake the GPS and Compass in INAV via MSP in order to use our external positioning system instead.

[Here](https://github.com/iNavFlight/inav/blob/master/src/main/msp/msp_protocol_v2_sensor.h) we found the corresponding MSP codes:

```C
#define MSP2_SENSOR_COMPASS         0x1F04
#define MSP2_SENSOR_GPS             0x1F03
```

And [here](https://github.com/iNavFlight/inav/blob/master/src/main/msp/msp_protocol_v2_sensor_msg.h) we found the structures of the corresponding messages:
```C
typedef struct __attribute__((packed)) {
    uint8_t  instance;
    uint32_t timeMs;
    int16_t  magX; // mGauss, front
    int16_t  magY; // mGauss, right
    int16_t  magZ; // mGauss, down
} mspSensorCompassDataMessage_t;

typedef struct __attribute__((packed)) {
    uint8_t  instance;                  // sensor instance number to support multi-sensor setups
    uint16_t gpsWeek;                   // GPS week, 0xFFFF if not available
    uint32_t msTOW;
    uint8_t  fixType;
    uint8_t  satellitesInView;
    uint16_t horizontalPosAccuracy;     // [cm]
    uint16_t verticalPosAccuracy;       // [cm]
    uint16_t horizontalVelAccuracy;     // [cm/s]
    uint16_t hdop;
    int32_t  longitude;
    int32_t  latitude;
    int32_t  mslAltitude;       // cm
    int32_t  nedVelNorth;       // cm/s
    int32_t  nedVelEast;
    int32_t  nedVelDown;
    uint16_t groundCourse;      // deg * 100, 0..36000
    uint16_t trueYaw;           // deg * 100, values of 0..36000 are valid. 65535 = no data available
    uint16_t year;
    uint8_t  month;
    uint8_t  day;
    uint8_t  hour;
    uint8_t  min;
    uint8_t  sec;
} mspSensorGpsDataMessage_t;
```

My questions are:
- What do all these fields expect exactly?
- Do I really need to fake all these fields in order to fake the GPS?

Essentially, my problem is that I am in an Euclidean basis of which I know, say, the longitude, latitude etc. of the origin, and I am trying to convert my Euclidean coordinates into GPS and Compass coordinates so that I can use NAV_POSHOLD with my external positioning system.

Thanks!

**Comment by lida2003** [comment]
I'm interested in this topic also. Is there any fake GPS unit/modules avaliable, such as UWB inhouse position system or others, that can be used as a test.

**Reply by yannbouteiller** [reply]
Hi,I finally figured it out, you can find the code [here](https://github.com/thecognifly/cognifly-python/blob/69494eae39320051e0228de6f0ef94f404c8cecc/cognifly/cognifly_controller/cognifly_controller.py#L343) (basically we fake everything that is required for NAV_POSHOLD)

**Reply by lida2003** [reply]
Hummm... so it's for simulation of NAV_POSHOLD or is there a sensor to fake position data and send to FC by MSP GPS protocol?

**Reply by yannbouteiller** [reply]
There is an external sensor, which we use to fake everything via MSP.

**Reply by lida2003** [reply]
That's a great job. 
I supposed that u have integrated many sensors to Raspeberry Pi. So you can fake any info. that iNav(FC)  need.
May I know what kind of sensors you have equiped? 

Any links  is well appreciated.

**Reply by yannbouteiller** [reply]
We use a visual odometry camera (t265) but the code is designed such that people can use any similar external ground truth pose estimator (e.g. Optitrack, UWB...). It can also be mixed with real sensors (by activating these sensors in INAV 6 and setting their "custom" argument to False on instantiation of the CogniflyController object).

**Reply by lida2003** [reply]
I read the CogniFly Project (hope to know more), does your drone uses following key componenets?

a) Google AIY Vision Kit (Raspberry Pi Zero WH +  Raspberry Pi Camera v2 )
b) odometry camera(t265)
c) Flight Controller (for iNav or BF)
d) bioinspired exoskeleton quard frame

It looks like Raspberry Pi Zero WH is in charge, which takes companion computer's position, handling XYZ+PRYT sensor data and do video AI analysis.

Is there any other sensor data fusion?
a) FC baro data disabled?
b) NO GPS, of course
c) optical flow, don't needed
d) Lidar or sonar or microwave sensor?

And is the [cognifly-python](https://github.com/thecognifly/cognifly-python) decision maker, brain?

**Reply by yannbouteiller** [reply]
We have several versions of the drone, either using the Raspberry Pi + RPI Camera, or the Google Coral board more more serious on-board AI applications. @ricardodeazambuja has done some stuff with the AIY vision kit.

Indeed, the drone uses a "bio-inspired" flexible exoskeleton that enables it hitting things / crashing without breaking, and has INAV running on a flight controller (the RPI / Coral communicates with the FC via MSP through a serial port), and in our current experiments we use a t265 to fake everything (GPS/barometer/compass) in order to fly indoor, which works nicely. 

cognifly-python is the python library that enables us to control the drone in a programmatical fashion, and do this sensor faking using the current master branch of INAV 6. This sensor faking is only implemented on the yb/stable branch of the repository at the moment, I'll work on a couple features before merging it with the main branch and updating the library on PyPI, as this library is used by students in robotic courses and I cannot update it too carelessly. The yb/stable branch works stably with our t265 custom estimator and sensor faking, though.

**Reply by lida2003** [reply]
Do u have any video about posHold & altHold with t265 to fake everything (GPS/barometer/compass) flying indoor, which works nicely.

Since we are doing something much more like tello. We hope to add an AirUnit + FC(iNav) + sensors (optical flow + lidar) to achieve posHold+altHold and  roughtly x centimeter forward/backward/left/right.

PS: AirUnit with or without camera versions.

**Reply by ricardodeazambuja** [reply]
I am jumping in this discussion because our system is not working (yet) using the current master branch of INAV because our pull request #8534, that fixes a problem with using a MSP based barometer, is still waiting to be accepted / merged. Please, @yannbouteiller and @lida2003, make some comments, put a :+1: in the pull request to see if it gets accepted :smile:
